docs/NFF-0.1-spec.norg

```diff
@@ -0,0 +1,778 @@
+    A _Norg_ file is made up of /characters/.
+    A <character> is any Unicode [code point].
+
+    [code point]{https://en.wikipedia.org/wiki/Code_point}
+
+**** Whitespace
+     A {> character} is considered to be *whitespace* if it is any of the following:
+     - A tab `U+0009`
+     - A line feed `U+000A`
+     - A form feed `U+000C`
+     - A carriage return `U+000D`
+     - Any code point in the [Unicode Zs general category] (includes the normal space `U+0020`)
+
+     Any combination of the above is also considered whitespace.
+
+     Tabs are not expanded to spaces and since whitespace has no semantic meaning there is no need
+     to define a default tab stop. However, if a parser must (for implementation reasons) define a
+     tab stop, we suggest to set it to 4 spaces.
+
+     However, the following whitespace combinations are a <line ending>:
+     - A single line feed
+     - A single carriage return
+     - A carriage return immediately followed by a line feed
+
+     ^ Blank lines are irrelevant in _Norg_ because standalone whitespace has no semantic relevance.
+
+     [Unicode Zs general category]{https://www.fileformat.info/info/unicode/category/Zs/list.htm}
+
+**** Punctuation
+     A {> character} is considered to be *punctuation* if it is any of the following:
+     - A standard ASCII punctuation character: |`!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~`|
+
+       +TODO: check whether we want to support every punctuation mark from the `en_US.utf8` locale+
+
+**** Escaping
+     A single {> character} can be escaped if it is immediately preceded by a backslash, `\`
+     (`U+005C`). Any {> character} may be escaped /apart from/ {> character}s within verbatim
+     segments +TODO: add link to verbatim segments+.
+
+**** Special Character
+     - | | TODO: formalize this section
+     - Emojis
+     +- Should we support [entity references]?
+     [entity references]{https://spec.commonmark.org/0.30/#entity-references}
+     vhyrro: I'm personally very against this idea tbh. I don't see an immediate use case for it,
+     perhaps if it's a lifesaver for some we could modify the syntax a little by enforcing
+     &nbsp& instead of &nbsp;+
+
+*** Words
+    The _Norg_ format is designed to be parsed on a word-by-word basis from left-to-right through
+    the entire document /in a single pass/. This is possible because the language is [Free-form],
+    meaning that whitespace has no semantic meaning, and because the markup follows strict rules
+    which are outlined in the later sections of this document.
+
+    A *word* is considered to be any combination of {> character}s which are neither {# whitespace},
+    {# punctuation} nor other {# special character}s.
+
+*** Paragraph Segments
+    {# Words} are first combined into {# paragraph segments}. A segment gets terminated by any of
+    the following:
+    - A {> line ending}
+    - Any of the {# attached modifiers} +NOTE(vhyrro): Don't paragraph segments contain attached mods?+
```

Comment:
Yes, you are right. Same with linkables actually.. I guess that means `paragraph_segment` is actually a line, right? :sweat_smile:

Reply:
Yeah essentially a `paragraph_segment` is just a line of text :P

I can rephrase this to be a little more accurate.